src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingController.java

```diff
@@ -345,15 +346,38 @@ public void onChatMessage(ChatMessageEvent event) {
   }
 
   private synchronized void renderQueues() {
+    controllers.clear();
     List<MatchmakerQueueBean> queues = new ArrayList<>(teamMatchmakingService.getQueues());
     queues.sort(Comparator.comparing(MatchmakerQueueBean::getId));
     int queuesPerRow = Math.min(queues.size(), 4);
     List<VBox> queueCards = queues.stream().map(queue -> {
       MatchmakingQueueItemController controller = uiService.loadFxml("theme/play/teammatchmaking/matchmaking_queue_card.fxml");
       controller.setQueue(queue);
+      controller.setTeamMatchmakingController(this);
+      controllers.add(controller);
       controller.getRoot().prefWidthProperty().bind(queuePane.widthProperty().divide(queuesPerRow).subtract(queuePane.getHgap()));
       return controller.getRoot();
     }).collect(Collectors.toList());
     JavaFxUtil.runLater(() -> queuePane.getChildren().setAll(queueCards));
   }
+
+  public void onSearchButtonClicked() {
+    if (searchButton.isSelected()) {
+      controllers.forEach(controller -> {
+        if (controller.joinLeaveQueueButton.isSelected()) {
+          teamMatchmakingService.joinQueue(controller.getQueue()).thenAccept(success -> {
+            if (!success) {
+              searchButton.setSelected(false);
+            }
+          });
+        }
+      });
+    } else {
+      controllers.forEach(controller -> {
+        if (controller.joinLeaveQueueButton.isSelected()) {
+          teamMatchmakingService.leaveQueue(controller.getQueue());
+        }
+      });
```

Comment:
Rather than using the controllers for this tracking it would be better if there was a separate property of either a queue or a list of preferences that were used. Then we could save and reload it on start like we do for factions.

Reply:
Then this whole operation could be a function in the teamMatchmakingService rather than the responsibility of the controller

Reply:
good idea